# <p style="text-align: center;"> Predicting Diabetes and Recommending Food For Farm Workers </p>

**Abstract:**

This project aims to decrease the number of pre-diabetics/diabetics in the population of farmer workers. As a result, we decided to work on an AI application that can help estimate prediabetic based on data provided by Dr. Angelos Sikalidis. In this project, we also built a recommendation system, using an AI, to recommend food based on the measurement of vitamin levels given to the recommender system. We used different libraries such as NumPy, pandas, scikit-learn, and Keras library in this project. 

In [1]:
#import required libraries
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
import numpy as np

Let's load in the CSV. This file is the mock data of different measurements that will be taken of individual farm workers.
The measurements are:
- BMI
- Weight (Kg)
- Waist Circumfrence (cm)
- Systolic BP (mmHg)
- Diastolic BP (mmHg)
- HDL-c (mg/dl)
- LDL-c (mg/dl)
- Total Cholesterol (mg/dL)
- Atherogenicity index
- TAG (mg/dL)
- FPG (mg/dL)
- HbA1c (%)
- CRP (mg/L)
- Pulse Oximetry (%)
- Servings Grain (oz.)
- Servings Fruit (2c.)
- Servings Veg (2.5c.)
- Servings Pro (5.5oz.)
- Servings Dairy (1c.)
- Total Caloric Intake (kcal)
- % Energy CHO
- % Energy Pro
- % Energy Fat
- Vitamin D (mcg)
- Vitamin E (mg)
- Vitamin A (mcg)
- Vitamin K (mcg)
- Vitamin B12 (mcg)
- Folate (mcg)
- Category [Malnourished, Normal, Borderline Risk, Overweight At Risk, High Risk]
- Diabetic [Non-diabetic, pre-diabetic, diabetic]

First thing we want to do is classify each individuals into certain health categories. These health categories include:
- **Malnourished: suffering from malnutrition.**
- **Normal: conforming to a standard health.**
- **Borderline Risk: close to bad health.**
- **Overweight At Risk: above a weight considered normal or desirable.**
- **High Risk: very dangerous health**

In [19]:
df = pd.read_csv('mock_data.csv')
properties = list(df.columns.values)
properties.remove('Category')

y = df['Category']
c = [p for p in range(len(y))]
    
m_data = pd.DataFrame(np.zeros((len(y),len(properties))),columns=properties,index=c)

for i in range(df.shape[0]):
    for p in properties:
        data = df.iloc[i][p]
        if type(data) == str:
            print("We got a string, should not happen")
            data = np.float64(data.replace(',',''))
            if np.isnan(data):
                #outcome = df.iloc[i]['Category']
                cleanData(m_data,p,outcome,i)
            else:
                m_data.iloc[i][p] = data
                print("-Added to m_data")
        elif np.isnan(data):
            #outcome = df.iloc[i]['Category']
            pass
        else:
            m_data.iloc[i][p] = data

X = m_data[properties]
X.head

<bound method NDFrame.head of       BMI  Weight (Kg)  Waist Circumfrence (cm)  Systolic BP (mmHg)  \
0    17.0         50.0                     70.0                55.0   
1    18.0         54.0                     73.0                60.0   
2    16.0         56.0                     72.0                59.0   
3    17.0         57.0                     71.0                57.0   
4    18.0         53.0                     70.0                59.0   
..    ...          ...                      ...                 ...   
114  50.0        106.0                    105.0               182.0   
115  47.0        119.0                    106.0               181.0   
116  45.0        107.0                    107.0              1184.0   
117  47.0        118.0                    109.0               185.0   
118  50.0        120.0                    110.0               184.0   

     Diastolic BP (mmHg)  HDL-c (mg/dl)  LDL-c (mg/dl)  \
0                   66.0           27.0           55.0   
1

In [20]:
# Once we load in the file, lets save 20% of our data for testing.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=0)
y_train = keras.utils.to_categorical(y_train) #turning targets into categorical vectors training
y_test = keras.utils.to_categorical(y_test) #turning targets into categorical vectors testing
num_classes = len(y_test[0])

feature_vector_length = len(properties)
input_shape = (feature_vector_length,)

In [21]:
# Configure the model
c_model = keras.Sequential()
c_model.add(keras.layers.Dense(32,input_shape=input_shape,activation=tf.nn.relu,kernel_initializer='he_uniform'))
c_model.add(keras.layers.Dense(32,activation=tf.nn.relu,kernel_initializer='he_uniform'))
c_model.add(keras.layers.Dense(num_classes,activation=tf.nn.softmax))

# Compile the model
c_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
c_model.fit(X_train,y_train,epochs=64,verbose=1,validation_split=0.2)

Epoch 1/64
3/3 [==============================] - 0s 78ms/step - loss: 363.8148 - accuracy: 0.0395 - val_loss: 329.6237 - val_accuracy: 0.0526
Epoch 2/64
3/3 [==============================] - 0s 11ms/step - loss: 285.4069 - accuracy: 0.0658 - val_loss: 256.3740 - val_accuracy: 0.2105
Epoch 3/64
3/3 [==============================] - 0s 12ms/step - loss: 221.8641 - accuracy: 0.1842 - val_loss: 180.4756 - val_accuracy: 0.1053
Epoch 4/64
3/3 [==============================] - 0s 10ms/step - loss: 160.6728 - accuracy: 0.0395 - val_loss: 108.8998 - val_accuracy: 0.0000e+00
Epoch 5/64
3/3 [==============================] - 0s 14ms/step - loss: 94.1497 - accuracy: 0.2105 - val_loss: 68.6903 - val_accuracy: 0.5789
Epoch 6/64
3/3 [==============================] - 0s 13ms/step - loss: 70.1868 - accuracy: 0.4868 - val_loss: 63.3008 - val_accuracy: 0.4737
Epoch 7/64
3/3 [==============================] - 0s 11ms/step - loss: 66.2885 - accuracy: 0.4474 - val_loss: 53.9577 - val_accuracy: 0.6316
E

3/3 [==============================] - 0s 12ms/step - loss: 0.6066 - accuracy: 0.9211 - val_loss: 0.0072 - val_accuracy: 1.0000
Epoch 60/64
3/3 [==============================] - 0s 11ms/step - loss: 0.6205 - accuracy: 0.8947 - val_loss: 0.4372 - val_accuracy: 0.8947
Epoch 61/64
3/3 [==============================] - 0s 11ms/step - loss: 0.6822 - accuracy: 0.9079 - val_loss: 0.0092 - val_accuracy: 1.0000
Epoch 62/64
3/3 [==============================] - 0s 11ms/step - loss: 0.5627 - accuracy: 0.9079 - val_loss: 0.1883 - val_accuracy: 0.8947
Epoch 63/64
3/3 [==============================] - 0s 12ms/step - loss: 0.6514 - accuracy: 0.9211 - val_loss: 0.3326 - val_accuracy: 0.8947
Epoch 64/64
3/3 [==============================] - 0s 11ms/step - loss: 0.5279 - accuracy: 0.9211 - val_loss: 0.1706 - val_accuracy: 0.8947


In [22]:
test_results = c_model.evaluate(X_test,y_test,verbose=1)
print(f'Test results - Loss: {test_results[0]} - Accuracy: {test_results[1]*100}%')

# Save the model if it is 100% accurate
if test_results[1]*100 == 100:
    c_model.save('models')

1/1 [==============================] - 0s 1ms/step - loss: 0.1799 - accuracy: 0.9583
Test results - Loss: 0.17991043627262115 - Accuracy: 95.83333134651184%


This model that we trained gives us an accuracy of **[INSERT ACCURACY HERE]%**. Let's see what our model predicts for other data not used in the mock data.

Our farmer has the following:

[BMI,

Weight (Kg),

Waist Circumfrence (cm),

Systolic BP (mmHg),

Diastolic BP (mmHg),

HDL-c (mg/dl),

LDL-c (mg/dl),

Total Cholesterol (mg/dL),

Atherogenicity index,

TAG (mg/dL),

FPG (mg/dL),

HbA1c (%),

CRP (mg/L),

Pulse Oximetry (%),

Servings Grain (oz.),

Servings Fruit (2c.),

Servings Veg (2.5c.),

Servings Pro (5.5oz.),

Servings Dairy (1c.),

Total Caloric Intake (kcal),

% Energy CHO,

% Energy Pro,

% Energy Fat,

Vitamin D (mcg),

Vitamin E (mg),

Vitamin A (mcg),

Vitamin K (mcg),

Vitamin B12 (mcg),

Folate (mcg),

Diabetic]

**Farmer 0: [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0,0]**

This person is not diabetic but is **malnourished**. 

**Farmer 1: [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1,0]**

This person is not diabetic and has **normal health status**. 

**Farmer 2: [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2,1]**

This person is not diabetic and has **borderline risk health status**. 

**Farmer 3: [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3,2]**

This person is diabetic and is **overweight at risk health status**.

**Farmer 4: [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4,2]**

This person is also diabetic and is **high risk health status**

In [23]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]
farmer_01 = [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,0]
farmer_02 = [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,1]
farmer_03 = [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,2]
farmer_04 = [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,2]

t = [farmer_00,farmer_01,farmer_02,farmer_03,farmer_04]
res = c_model.predict(t)

for index,result in enumerate(res):
    print("Farmer ",index)
    print(res[index])
    print("-----------------------------------------------------------------------")

Farmer  0
[1.0000000e+00 1.9666030e-30 3.7964523e-25 0.0000000e+00 1.9370336e-26]
-----------------------------------------------------------------------
Farmer  1
[1.3343860e-15 1.0000000e+00 6.2096133e-24 0.0000000e+00 1.8725350e-29]
-----------------------------------------------------------------------
Farmer  2
[9.9238045e-31 3.5340985e-38 9.9983573e-01 1.6424965e-04 0.0000000e+00]
-----------------------------------------------------------------------
Farmer  3
[0.0000000e+00 0.0000000e+00 1.3066661e-30 1.0000000e+00 0.0000000e+00]
-----------------------------------------------------------------------
Farmer  4
[0. 0. 0. 0. 1.]
-----------------------------------------------------------------------


The output of our model is an array of probability for each indices 0-4.

Each index is a class for a health category.
- 0: Malnourished
- 1: Normal Health Status
- 2: Boderline Risk Health Status
- 3: Overweight at Risk Health Status
- 4: High Risk Health Status

At each indices, the probabilty tells us how likely it is that the farmer falls into the above category. Looking at famer 0:

Farmer  0: [1.0000000e+00 , 1.9666030e-30 , 3.7964523e-25 , 0.0000000e+00 , 1.9370336e-26]

At index 0, we have 1.0000000e+00 which is a probability of 100%.

At index 1, we have 1.9666030e-30 which is a probability of ~0%.

At index 2, we have 3.7964523e-25 which is a probability of ~0%.

At index 3, we have 0.0000000e+00 which is a probability of 0%.

At index 4, we have 1.9370336e-26 which is a probability of ~0%.

The model predicts that the farmer falls into index 0 category which is Malnourished, the correct health status

In [24]:
for i,rr in enumerate(res):
    max_i = 0
    max_r = 0
    print("Farmer ",i)
    if i == 0:
        print("True= Malnourished")
    elif i == 1:
        print("True= Normal")
    elif i == 2:
        print("True= Borderline Risk")
    elif i == 3:
        print("True= Overweight At Risk")
    elif i == 4:
        print("True= High Risk")
    for ri,r in enumerate(rr):
        if r > max_r:
            max_i = ri
            max_r = r
    if max_i == 0:
        print("Pred= Malnourished")
    elif max_i == 1:
        print("Pred= Normal")
    elif max_i == 2:
        print("Pred= Borderline Risk")
    elif max_i == 3:
        print("Pred= Overweight At Risk")
    elif max_i == 4:
        print("Pred= High Risk")
    print("-------------------------")

Farmer  0
True= Malnourished
Pred= Malnourished
-------------------------
Farmer  1
True= Normal
Pred= Normal
-------------------------
Farmer  2
True= Borderline Risk
Pred= Borderline Risk
-------------------------
Farmer  3
True= Overweight At Risk
Pred= Overweight At Risk
-------------------------
Farmer  4
True= High Risk
Pred= High Risk
-------------------------


From the output of our model compared to the true value of our data, for all the tests, the prediction of the model matches our data. However, if we retrain our model many times again, we get very different accuracies ranging from 60% accurate to 100% accuarate. We still cannot say that our model is 100% accurate because our data size is only 119. A much bigger  data will give us a more accutate result. 

Let's now categorize our mock data.

In [26]:
health_status = []

malnourished = []
maln_index = 0
normal = []
norm_index = 1
borderline_risk = []
border_index = 2
overweight_risk = []
overw_index = 3
high_risk = []
high_index = 4

res = c_model.predict(X)

for i,rr in enumerate(res):
    max_i = 0
    max_r = 0
    for ri,r in enumerate(rr):
        if r > max_r:
            max_i = ri
            max_r = r
    if max_i == 0:
        malnourished.append(list(X.loc[i]))
    elif max_i == 1:
        normal.append(list(X.loc[i]))
    elif max_i == 2:
        borderline_risk.append(list(X.loc[i]))
    elif max_i == 3:
        overweight_risk.append(list(X.loc[i]))
    elif max_i == 4:
        high_risk.append(list(X.loc[i]))
        
health_status.append(malnourished)
health_status.append(normal)
health_status.append(borderline_risk)
health_status.append(overweight_risk)
health_status.append(high_risk)

## Prediabetic or Diabetic ##

Now that we can classify each individuals into certain health categories, it would be very useful to know if these individuals are diabetic, pre-diabetic or not diabetic. This way when we recommend food to a person, we can be sure we're not recommending diabetics a certain food they aren't allowed to consume. 

In [31]:
properties = list(df.columns.values)
properties.remove('Diabetic')

y_diabetes = df['Diabetic']

In [32]:
# Splitting test and training sets
Xdiabetes_train, Xdiabetes_test, ydiabetes_train, ydiabetes_test = train_test_split(X,y_diabetes,test_size=0.2,random_state=0)
ydiabetes_train = keras.utils.to_categorical(ydiabetes_train) #turning targets into categorical vectors training
ydiabetes_test = keras.utils.to_categorical(ydiabetes_test) #turning targets into categorical vectors testing
num_classes = len(ydiabetes_test[0])

feature_vector_length = len(properties)
input_shape = (feature_vector_length,)

In [33]:
diabetes_model = keras.Sequential()
diabetes_model.add(keras.layers.Dense(32,input_shape=input_shape,activation=tf.nn.relu,kernel_initializer='he_uniform'))
diabetes_model.add(keras.layers.Dense(32,activation=tf.nn.relu,kernel_initializer='he_uniform'))
diabetes_model.add(keras.layers.Dense(num_classes,activation=tf.nn.softmax))

diabetes_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])
history_diabetes = diabetes_model.fit(Xdiabetes_train,ydiabetes_train,epochs=64,verbose=1,validation_split=0.2)

Epoch 1/64
3/3 [==============================] - 0s 67ms/step - loss: 292.5675 - accuracy: 0.2763 - val_loss: 214.8682 - val_accuracy: 0.3684
Epoch 2/64
3/3 [==============================] - 0s 14ms/step - loss: 209.0755 - accuracy: 0.1579 - val_loss: 149.9653 - val_accuracy: 0.0526
Epoch 3/64
3/3 [==============================] - 0s 17ms/step - loss: 142.9689 - accuracy: 0.1053 - val_loss: 104.2991 - val_accuracy: 0.3684
Epoch 4/64
3/3 [==============================] - 0s 16ms/step - loss: 107.7402 - accuracy: 0.2895 - val_loss: 83.3034 - val_accuracy: 0.3684
Epoch 5/64
3/3 [==============================] - 0s 13ms/step - loss: 86.7231 - accuracy: 0.2500 - val_loss: 64.9096 - val_accuracy: 0.2632
Epoch 6/64
3/3 [==============================] - 0s 13ms/step - loss: 65.1750 - accuracy: 0.2763 - val_loss: 41.3122 - val_accuracy: 0.4737
Epoch 7/64
3/3 [==============================] - 0s 14ms/step - loss: 41.9929 - accuracy: 0.4211 - val_loss: 16.5693 - val_accuracy: 0.4211
Epoch 

3/3 [==============================] - 0s 18ms/step - loss: 1.9145 - accuracy: 0.8026 - val_loss: 0.9536 - val_accuracy: 0.8421
Epoch 60/64
3/3 [==============================] - 0s 60ms/step - loss: 1.3875 - accuracy: 0.8289 - val_loss: 0.4363 - val_accuracy: 0.9474
Epoch 61/64
3/3 [==============================] - 0s 68ms/step - loss: 1.3548 - accuracy: 0.8816 - val_loss: 0.9183 - val_accuracy: 0.8421
Epoch 62/64
3/3 [==============================] - 0s 74ms/step - loss: 1.3253 - accuracy: 0.8947 - val_loss: 0.4271 - val_accuracy: 0.9474
Epoch 63/64
3/3 [==============================] - 0s 44ms/step - loss: 1.4285 - accuracy: 0.8816 - val_loss: 0.4087 - val_accuracy: 0.8947
Epoch 64/64
3/3 [==============================] - 0s 17ms/step - loss: 1.0710 - accuracy: 0.9211 - val_loss: 0.3878 - val_accuracy: 0.8947


In [35]:
test_results_diabetes = diabetes_model.evaluate(Xdiabetes_test,ydiabetes_test,verbose=1)
print(f'Test results - Loss: {test_results_diabetes[0]} - Accuracy: {test_results_diabetes[1]*100}%')

# Save the model if it is 100% accurate
if test_results_diabetes[1]*100 == 100:
    diabetes_model.save('models')

1/1 [==============================] - 0s 2ms/step - loss: 0.0379 - accuracy: 1.0000
Test results - Loss: 0.03787187859416008 - Accuracy: 100.0%
INFO:tensorflow:Assets written to: models/assets


Using the same farmers as before, lets see what our model predicts.

**Farmer 0: [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]**

This person is **not diabetic** but is malnourished

**Farmer 1: [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1]**

This person is **not diabetic** and has normal health status.

**Farmer 2: [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2]**

This person is **not diabetic** and has borderline risk health status

**Farmer 3: [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3]**

This person is **diabetic** and is overweight at risk health status

**Farmer 4: [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4]**

This person is also **diabetic** and is high risk health status


In [36]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]
farmer_01 = [24,63,81,65,69,76,75,169,0,109,86,4.2,0.53,102,3,2.5,2.5,2,1.5,1800,52,34,19,55,12,999,124,2,415,1]
farmer_02 = [28,85,84,130,84,57,100,192,0.21,200,111,6.3,1.8,90,7,0,0,0,0,2300,59,40,39,0,0,0,0,0,0,2]
farmer_03 = [30,92,91,165,99,42,158,208,0.25,398,124,6.5,2.8,89,9,0,0,7,0,2600,62,40,40,0,0,0,0,0,0,3]
farmer_04 = [45,106,105,182,130,31,176,256,0.38,520,130,9.5,3.8,80,12,0,0,9,0,3400,69,48,51,101,24,3025,147,3.4,640,4]

res_diabetic = diabetes_model.predict([farmer_00,farmer_01,farmer_02,farmer_03,farmer_04])
#print(res_diabetic)

for i,result in enumerate(res_diabetic):
    print("Farmer ",i)
    max_r = 0
    max_i = 0
    for ir,r in enumerate(result):
        if r > max_r:
            max_r = r
            max_i = ir
    if i == 0:
        print("True: Not diabetic")
    elif i == 1:
        print("True: Not diabetic")
    elif i == 2:
        print("True: Not diabetic")
    elif i == 3:
        print("True: Diabetic")
    elif i == 4:
        print("True: Diabetic")
    if max_i == 0:
        print("Pred: Not diabetic")
    elif max_i == 1:
        print("Pred: Prediabetic")
    else:
        print("Pred: Diabetic")
    print("------------------")


Farmer  0
True: Not diabetic
Pred: Not diabetic
------------------
Farmer  1
True: Not diabetic
Pred: Prediabetic
------------------
Farmer  2
True: Not diabetic
Pred: Diabetic
------------------
Farmer  3
True: Diabetic
Pred: Diabetic
------------------
Farmer  4
True: Diabetic
Pred: Diabetic
------------------


Even though our model is 100% accurate, we are still getting wrong predictions. A bigger sample size would help make the model more accurate.

## Recommendation System

**A very basaic CSV file that contains about 7,412 food variety and their nutritional facts**

In [37]:
#import external libraries and functions
import sklearn
from sklearn.neighbors import NearestNeighbors

#load dataset from CSV file and show first 5 records
food_bank = pd.read_csv('food.csv')
   
food_bank.columns = food_bank.columns
food_bank.head()

Category               Description  Nutrient Data Bank Number  \
0      BUTTER          BUTTER,WITH SALT                       1001   
1      BUTTER  BUTTER,WHIPPED,WITH SALT                       1002   
2  BUTTER OIL      BUTTER OIL,ANHYDROUS                       1003   
3      CHEESE               CHEESE,BLUE                       1004   
4      CHEESE              CHEESE,BRICK                       1005   

   Data.Alpha Carotene  Data.Ash  Data.Beta Carotene  Data.Beta Cryptoxanthin  \
0                    0      2.11                 158                        0   
1                    0      2.11                 158                        0   
2                    0      0.00                 193                        0   
3                    0      5.11                  74                        0   
4                    0      3.18                  76                        0   

   Data.Carbohydrate  Data.Cholesterol  Data.Choline  ...  \
0               0.06               215            19  ...   
1               0.06               219            19  ...   
2               0.00               256            22  ...   
3               2.34                75            15  ...   
4               2.79                94            15  ...   

   Data.Major Minerals.Potassium  Data.Major Minerals.Sodium  \
0                             24                         576   
1                             26                         827   
2                              5                           2   
3                            256                        1395   
4                            136                         560   

   Data.Major Minerals.Zinc  Data.Vitamins.Vitamin A - IU  \
0                      0.09                          2499   
1                      0.05                          2499   
2                      0.01                          3069   
3                      2.66                           763   
4                      2.60                          1080   

   Data.Vitamins.Vitamin A - RAE  Data.Vitamins.Vitamin B12  \
0                            684                       0.17   
1                            684                       0.13   
2                            840                       0.01   
3                            198                       1.22   
4                            292                       1.26   

   Data.Vitamins.Vitamin B6  Data.Vitamins.Vitamin C  Data.Vitamins.Vitamin E  \
0                     0.003                      0.0                     2.32   
1                     0.003                      0.0                     2.32   
2                     0.001                      0.0                     2.80   
3                     0.166                      0.0                     0.25   
4                     0.065                      0.0                     0.26   

   Data.Vitamins.Vitamin K  
0                      7.0  
1                      7.0  
2                      8.6  
3                      2.4  
4                      2.5  

[5 rows x 48 columns]

In [38]:
data_key = pd.read_csv('mock_data_key.csv')

print(data_key)

                            MALNOURISHED NORMAL/DESIRABLE BORDERLINE RISK  \
BMI                                 < 19            19-25           26-29   
Weight (Kg)                         < 63            63-68           69-86   
Waist Circumference (cm)            ≤ 73            74-81           82-89   
Systolic BP (mmHg)                  ≤ 60           61-129         130-139   
Diastolic BP (mmHg)                 ≤ 67            68-79           80-89   
HDL-c (mg/dL)                       < 30           60-100           46-59   
LDL-c (mg/dL)                       < 65            65-90          91-130   
Total Cholesterol (mg/dL)          < 110          160-180         181-200   
Atherogenicity index                   X         −0.3-0.1       0.11-0.24   
TAG (mg/dL)                            X            < 150         150-200   
FPG (mg/dL)                         < 75           75-100         100-115   
HbA1c (%)                            < 4            4-5.7         5.7-6.4   

**The data_key tells us where each measurements of our features needs to be at in order to fall under certain health categories.**

Lets try to recommend a food to this person: 

farmer_00 = [17,
            50,
            70,
            55,
            66,
            27,
            55,
            100,
            0,
            0,
            48,
            3.8,
            0,
            84,
            2,
            0,
            0,
            1,
            0,
            1799,
            40,
            20,
            15,
            25,
            **10 Vitamin E,
            699 Vitamin A,
            75 Vitamin K,
            1.4 Vitamin B12,**
            390,
            0]

This famer is categorized as malnourished and has 
- 10 mg of Vitamin E (mg)
- 699 mcg of Vitamin A (mcg)
- 75 mcg of Vitamin K (mcg)
- 1.4 mcg of Vitamin B12 (mcg)

The recommended amount:
- Vitamin E is 12-18 mg
- Vitamin A is 700-1000 mcg
- Vitamin K is 110-140  mcg
- Vitamin B12 is 1.5-3 mcg

We can find the difference of the recommended and actual to get the proper food recommendation.

Need:
- 2 mg of Vitamin E
- 1 mcg of Vitamin A
- 35 mcg of Vitamin K
- 0.1 mcg of Vitamin B12

Since we lack the most on Vitamin K and Vitamin B12, we need to recommend a food that will help normalize these levels.

In [39]:
farmer_00 = [17,50,70,55,66,27,55,100,0,0,48,3.8,0,84,2,0,0,1,0,1799,40,20,15,25,10,699,75,1.4,390,0]

# Indices to where each farmer's vitamin measures are
v_e = 24
v_a = 25
v_k = 26
v_b12 = 27

# Food bank indices of vitamins we want
f_a_index = 41
f_b12_index = 43
f_e_index = 46
f_k_index = 47

# Recommended Vitamins
rec_vitamins = {}
rec_vitamins["vitamin_e"] = 12
rec_vitamins["vitamin_a"] = 700
rec_vitamins["vitamin_k"] = 110
rec_vitamins["vitamin_b12"] = 1.5

recommendations = []

# Input: A list of vitamins that might be needed
# Outputs: A dictionary of the vitamin name as key and the amount of vitamin needed as value
def within_range(need_list):
    threshold = 5
    vitamins_needed = {}

    vitamins_needed["Vitamin_E"] = 0
    vitamins_needed["Vitamin_A"] = 0
    vitamins_needed["Vitamin_K"] = 0
    vitamins_needed["Vitamin_B12"] = 0
    
    # If a person only need 5 or less units of vitamin, we don't recommend anything
    for ni,n in enumerate(need_list):
        if n > threshold:
            vitamin = ""
            if ni == 0:
                vitamins_needed["Vitamin_E"] = n
            elif ni == 1:
                vitamins_needed["Vitamin_A"] = n
            elif ni == 2:
                vitamins_needed["Vitamin_K"] = n
            elif ni == 3:
                vitamins_needed["Vitamin_B12"] = n
    return vitamins_needed
    
# Input: 
# Output:
def get_food_rec(farmer):
    # Get the farmer's needed vitamins
    need_vitamin_e = rec_vitamins["vitamin_e"] - farmer[v_e]
    need_vitamin_a = rec_vitamins["vitamin_a"] - farmer[v_a]
    need_vitamin_k = rec_vitamins["vitamin_k"] - farmer[v_k]
    need_vitamin_b12 = rec_vitamins["vitamin_b12"] - farmer[v_b12]
    
    need_list = []
    need_list.append(need_vitamin_e)
    need_list.append(need_vitamin_a)
    need_list.append(need_vitamin_k)
    need_list.append(need_vitamin_b12)

    vitamins_needed = within_range(need_list)
    
    X_rec = food_bank.iloc[:,[f_e_index,f_a_index,f_k_index,f_b12_index]].values
    nbrs = NearestNeighbors(n_neighbors=1).fit(X_rec)
    
    foods = {}
    food_quantity = {}
    
    not_in_range = True
    for key in vitamins_needed:
        if vitamins_needed[key] != 0:
            food_list = []
            if key == "Vitamin_E":
                while vitamins_needed[key] > 5:
                    food = food_bank.iloc[nbrs.kneighbors([[vitamins_needed["Vitamin_E"],0,0,0]])[1][0][0]]
                    if food[0] not in food_quantity:
                        food_quantity[food[0]] = 1
                        food_list.append(food)
                    else:
                        food_quantity[food[0]] += 1
                    vitamins_needed[key] -= food[f_e_index]
                foods["Vitamin_E"] = food_list
            elif key == "Vitamin_A":
                while vitamins_needed[key] > 5:
                    food = food_bank.iloc[nbrs.kneighbors([[0,vitamins_needed["Vitamin_A"],0,0]])[1][0][0]]
                    if food[0] not in food_quantity:
                        food_quantity[food[0]] = 1
                        food_list.append(food)
                    else:
                        food_quantity[food[0]] += 1
                    vitamins_needed[key] -= food[f_a_index]
                foods["Vitamin_A"] = food_list
            elif key == "Vitamin_K":
                while vitamins_needed[key] > 5:
                    food = food_bank.iloc[nbrs.kneighbors([[0,0,vitamins_needed["Vitamin_K"],0]])[1][0][0]]
                    if food[0] not in food_quantity:
                        food_quantity[food[0]] = 1
                        food_list.append(food)
                    else:
                        food_quantity[food[0]] += 1
                    vitamins_needed[key] -= food[f_k_index]
                foods["Vitamin_K"] = food_list
            elif key == "Vitamin_B12":
                while vitamins_needed[key] > 5:
                    food = food_bank.iloc[nbrs.kneighbors([[0,0,0,vitamins_needed["Vitamin_B12"]]])[1][0][0]]
                    if food[0] not in food_quantity:
                        food_quantity[food[0]] = 1
                        food_list.append(food)
                    else:
                        food_quantity[food[0]] += 1
                    vitamins_needed[key] -= food[f_b12_index]
                foods["Vitamin_B12"] = food_list
    return foods,food_quantity

def print_rec(foods,food_quantity):
    for k in foods:
        print("For " + k + " we recommend:")
        for food in foods[k]:
            print(food[0] + " x" + str(food_quantity[food[0]]) + " serving(s)")
        print("-------------------------------")
        
foods,food_quantity = get_food_rec(farmer_00)
print_rec(foods,food_quantity)


For Vitamin_K we recommend:
CASHEW NUTS x1 serving(s)
-------------------------------


We got this food: **CASHEW NUTS (Vitamin K)**

For Cashew Nuts, we needed about 35 mcg more.

The information about it on the Food CVS gives us the following:

Data.Vitamins.Vitamin A:                                                         0

Data.Vitamins.Vitamin B12:                                                            0

Data.Vitamins.Vitamin E:                                                           0.92

**Data.Vitamins.Vitamin K:                                                           34.7**


What kind of food are recommended to farmer that is of normal health status?

In [40]:
n_farmer = health_status[norm_index][0]
n_food,food_q = get_food_rec(n_farmer)
print_rec(n_food,food_q)

This farmer is of normal health status and don't lack any vitamins, so we did not recommend any food for them.

## Future Works

One of the features that we would have liked to implement is having a recommendation system for physical activities. Similarly to a food recommendation, we would have an AI that recommends activities, for an individual, based on previously performed activities to stay healthy.
Another feature that could be implemented for this project is using an Apple watch. We could take advantage of all the sensors that the watch provides such as the heart rate monitor. Also, since Apple has announced an affordable watch, it would be a bit more accessible to our targeted user.
